### ISSUE TITLE
Make the deprecations as exceptions machinery more general

### ISSUE THREAD

**Issue by drdavella** [astropy/astropy#7933]
There's no reason that we can't provide an API that would allow other kinds of warnings to be treated as exceptions as well.

For example, I'm trying to track down the `ResourceWarning` that was reported in #7913. It would be really useful if I could just treat all `ResourceWarning`s as exceptions so that I could get more informative tracebacks.

If we make this change, then `treat_deprecations_as_exceptions` just becomes a special case and calls the more general function.

**Comment by jdavies-st** [comment]
There's functionality similar to this built into `pytest`:
```
pytest -W error::RuntimeWarning
```
will turn any runtime warnings into errors with full tracebacks (which can be controlled via pytest's `-tb` flag).  One can use any `Warning` subclass, though I haven't figured out how to do `ValidationWarning` for instance on json schema.

Does the `deprecations_as_exceptions()` machinery do more?

**Comment by jdavies-st** [comment]
Agreed.  One often has to work backwards.  I.e. convert all warnings to exceptions, then add extra `-W` flags for warnings one would like to ignore until you've narrowed it to the custom warning you're searching for.  And for a large test suite, this is pretty irritating.

**Comment by drdavella** [comment]
I agree with @Cadair and also think that there is some value in programmatically making this part of all test runs, although I guess that is also possible for the simple cases using the `addopts` argument of `[tool:pytest]` in `setup.cfg`.

**Comment by jdavies-st** [comment]
One way to get around the issue of using the `-W` flag with custom warning subclasses is to don't worry about the subclass and instead use the following form:
```
pytest -W "ignore: The method AsdfFile.open"
```
to filter out warnings such as
```
    asdf.py:738: AsdfDeprecationWarning: The method AsdfFile.open has been deprecated and will be removed in asdf-3.0. Use the top-level asdf.open function instead.
```
which filters not on the warning subclass (`AsdfDeprecationWarning` above), but on the warning message itself.

If you want to then get a full traceback, you can convert these warnings to errors as mentioned above:
```
pytest -W "error: The method AsdfFile.open"
```
